### DISCUSSION TITLE
How to define UMAT?

### DISCUSSION THREAD

**Discussion by cunyizju** [OpenRadioss/OpenRadioss#1164] [Q&A]
Hey, I am learning how to develop UMAT using OpenRadioss.
Based on the official document, I put LECM29.F in starter/source/mat/mat029 and sigeps.F in engine/source/materials/mat029.
These subroutines are used for elastic constitutive relations. When I compiled the source code and run the starter, the error msg was given below:
```
ERROR ID :   1130
 ** ERROR IN USER INTERFACE
DESCRIPTION :
   OPTION: /MAT/USER01
   USER INTERFACE OPTION USED, BUT NO USER INTERFACE ROUTINE IS FOUND
 RADIOSS STOP DUE TO INPUT ERROR
     ERROR TERMINATION
     ------------------
              1ERROR(S)
 ** ERROR: FILE elasticity_0000_0001.rst NOT FOUND
     ERROR TERMINATION
--------------------------------------------------------------------------
MPI_ABORT was invoked on rank 0 in communicator MPI_COMM_WORLD
with errorcode 2.

NOTE: invoking MPI_ABORT causes Open MPI to kill all MPI processes.
You may or may not see output from other processes, depending on
exactly when Open MPI kills them.
```
Could anyone let me know how to solve this problem? Many thanks in advance!

**Comment by MarianBulla** [comment]
Hey, can you double check it with the example, we add in the userlib_sdk folder, please ?
Thanks and best regards,
Marian

**Comment by OlivierWienholtz** [comment]
Hey Bruce,

Just a few additionnal comments to Marian, indeed, adding the code in Radioss & building Radioss  won't work.
 
Our User Interface UMAT is working as a Plugin with a dynamic Library. 
You need to generate a Linux .so library or a Windows .dll with your user code. 
This library will be loaded by Radioss & the code applied.

We have created an SDK to generate this Library (we named it user library / userlib : made of some Libraries + scripts).
This SDKK is now available on Github on the Tools Repository on the OpenRadioss organization : 
https://github.com/OpenRadioss/Tools

Please have a look at the Userlib_sdk repo :  how to build this SDK,  how to generate the library with your code,
and how to use with Radioss.

Another Note : 
You are using the older law29-31  interface. I would recommend to use the newer luser01-luser99 interface. 
You will have access to more Parameters from Radioss laws.

Hope it helps,
Best Regards,
Olivier

**Comment by OlivierWienholtz** [comment]
Bring your topic in the Discution Area.

**Reply by cunyizju** [reply]
Thanks @MarianBulla @OlivierWienholtz  for your suggestions.

I have generated 'libraduser_linux64_gfortran.a'. 
However, according to the following instructions, I could not obtain 'libraduser_linux64.so'.

```
$RAD_USERLIB_SDK_PATH/$RAD_USERLIB_ARCH/build_userlib.bash starter="lecmuser01.f engine="luser01c.f"

*********************************************
** Generating Radioss Dynamic user library **
**            linux64 gfortran             **
**                                         **
*********************************************



Compiling:  lecmuser01.f
-----------



Compiling:  luser01c.f
----------



Creating library:  libraduser_linux64.so
-----------------



Done
----
```

BTW, 
```
build_userlib.bash starter="lecmuser01.f engine="luser01c.f"
```
may be needed to modified as
```
. build_userlib.bash starter="lecmuser01.f" engine="luser01.f"
```

I am using wsl with Ubuntu 20.04. When I run 
```
export OPENRADIOSS_PATH=/home/bruce20/OpenSource/OpenRadioss
export RAD_CFG_PATH=$OPENRADIOSS_PATH/hm_cfg_files
export RAD_H3D_PATH=$OPENRADIOSS_PATH/extlib/h3d/lib/linux64
export OMP_STACKSIZE=400m
export LD_LIBRARY_PATH=$OPENRADIOSS_PATH/extlib/hm_reader/linux64/:$LD_LIBRARY_PATH

export LD_LIBRARY_PATH=/opt/openmpi/lib:$LD_LIBRARY_PATH
export PATH=/opt/openmpi/bin:$PATH

export OMP_STACKSIZE=400m
export OMP_NUM_THREADS=1

export RAD_USERLIB_SDK_PATH=/home/bruce20/OpenSource/Tools/userlib_sdk/
export RAD_USERLIB_ARCH=linux64_gfortran

. build_userlib.bash starter="lecmuser01.f" engine="luser01.f"
```
The command window quitted unexpectedly. Could you please help me solve this problems? Many thanks~

Kind regards,
Bruce

**Reply by MarianBulla** [reply]
Hi Bruce, 

Great to see, you were able to compile your own material model :-)

can you please check if you are on WSL version 1 ?
If so, please update WSL to version 2.

Hope it helps you ?
Thanks and best regards,
Marian

**Reply by cunyizju** [reply]
Thanks, Marian. It is wsl2.

I am not sure if the files were properly organised.
![1679386216104](https://user-images.githubusercontent.com/40728007/226549029-80ff15c1-8740-4e0a-8403-94d0d5d4c6c1.png)
It would be great to detect some bugs in this figure.

Best wishes,
Bruce

**Reply by MarianBulla** [reply]
Hi Bruce,
for beeing able to create your own USER subroutines, you need the SDK, which is in the Tools repository.
After creating the sdk, you need to create 2 environment variables (see screenshot).
Then you can strart building your own user materials.
![image](https://user-images.githubusercontent.com/105412502/226586853-b052d0be-65a6-4b31-afe5-0f02911506d2.png)

**Reply by MarianBulla** [reply]
Then, adding the environment variables for getting OpenRadioss running:

![image](https://user-images.githubusercontent.com/105412502/226587868-25b45f61-dcd5-4ba1-93a9-847fe4a1aba0.png)

**Reply by MarianBulla** [reply]
There is an example USER material for solids, within this repo.
You can try it with this one, first.

Hope it helps you ?

Thanks and best regards,
Marian
